### PR TITLE
Add bookmarks hooks

### PR DESCRIPTION
**Pull request by ashleemboyer**
## 🔗 Relevant links

- [Preview link](https://dev-portal-git-ambadd-bookmarks-hooks-hashicorp.vercel.app/) 🔎
- [Asana task](https://app.asana.com/0/1202097197789424/1202791375540709/f) 🎟️

## 🗒️ What

- Installs React Query and the Devtools
- Updates `_app.tsx` to work with React Query and the Devtools
- Adds a few new hooks:
  - `useAllBookmarks`
  - `useBookmarkMutations`
  - `useBookmarksByTutorialIds`
  - `useIsBookmarked`
- Updates `Tutorial['id']` to reference `ApiTutorial['id']`
- Updates some of the bookmark API methods to reference `Tutorial['id']` instead of `ApiTutorial['id']`
  - `createBookmark`
  - `deleteBookmark`
  - `getBookmark`
  - `getBookmarksByTutorialIds`
- Updates `ProfileView` to be a test page for rendering all bookmarks and some test tutorial cards
- Removes some no longer needed files for `ProfileView` (and they aren't needed in upcoming designs)

## 🧪 Testing

Throughout the steps below, have the React Query Devtools open and make sure the queries behave how you expect them to.

- [ ] Check out and run the code locally
- [ ] Go to /profile
- [ ] In a card in the Tutorials section of the page, click an "add bookmark" button
- [ ] The Bookmarks list should automatically update
- [ ] The "add bookmark" button you clicked should automatically update

### CODE REVIEW

File: src/components/tutorial-card/types.ts
Comment (kendallstrautman) [inline src/components/tutorial-card/types.ts:8]: This would actually be `Tutorial['id]` from this file, https://github.com/hashicorp/dev-portal/blob/main/src/lib/learn-client/types.ts#L11 - since this data is formatted. Just a string so not that it matters much.

File: src/components/bookmark-button/index.tsx
Comment (kendallstrautman) [inline src/components/bookmark-button/index.tsx:13]: Do you mind if we do this integration in another PR? Just to avoid conflicts with some refactoring work I'm doing [in this PR](https://github.com/hashicorp/dev-portal/pull/837)

Comment (kendallstrautman) [reply src/components/bookmark-button/index.tsx:13]: Once this is merged I can integrate these updates into that PR

Comment (ashleemboyer) [reply src/components/bookmark-button/index.tsx:13]: Sure

File: src/hooks/bookmarks/use-all-bookmarks.ts
Comment (kendallstrautman) [inline src/hooks/bookmarks/use-all-bookmarks.ts:7]: Would the data be 'undefined' if someone isn't authenticated? or what would cause that return

File: src/hooks/bookmarks/use-is-bookmarked.ts
Comment (kendallstrautman) [inline src/hooks/bookmarks/use-is-bookmarked.ts:20]: Hmmm I would expect us to just fetch once single tutorial here. Trying to think through this.

File: src/hooks/bookmarks/use-all-bookmarks.ts
Comment (ashleemboyer) [reply src/hooks/bookmarks/use-all-bookmarks.ts:7]: The data is `undefined` while the data is being loaded for the first time. So in the case that someone is not authenticated, the query will never execute and the value will be `undefined`.

File: src/hooks/bookmarks/use-is-bookmarked.ts
Comment (ashleemboyer) [reply src/hooks/bookmarks/use-is-bookmarked.ts:20]: Based on how this hook is intended to be used (once for each `TutorialCard` or `TutorialMeta`), that would require doing a fetch for each card/meta. So if there are 20 cards on a page, that means 20 fetches. This approach reduces it to one fetch in total.

File: src/components/tutorial-card/types.ts
Comment (ashleemboyer) [reply src/components/tutorial-card/types.ts]: Updated in https://github.com/hashicorp/dev-portal/pull/838/commits/f6816373e5e945e458540c30519570929b66f3cf. Also replaced `TutorialCardPropsWithId` since it's redundant with these overall changes.

File: src/components/bookmark-button/index.tsx
Comment (ashleemboyer) [reply src/components/bookmark-button/index.tsx:13]: Actually on second thought, I'm not sure these changes should be undone. They are being used to test the new hooks.

File: src/hooks/bookmarks/use-bookmarks-by-tutorial-ids.ts
Comment (kendallstrautman) [inline src/hooks/bookmarks/use-bookmarks-by-tutorial-ids.ts:18]: I was considering that the page level view would fetch the specific set of tutorial bookmark data with this function. And then pass that down to the cards, instead of each card calling the hook. 🤔 Not that we have to do that but it was just my thought. 

It does seem a bit odd to me having every tutorial card calling the 'getAllBookmarks' hook. But maybe with the caching that's okay?....

File: src/components/bookmark-button/index.tsx
Comment (kendallstrautman) [reply src/components/bookmark-button/index.tsx:13]: Once were done testing, we could undo? Or I can just manage the conflicts...

File: src/components/tutorial-card/types.ts
Comment (kendallstrautman) [reply src/components/tutorial-card/types.ts]: Cool, I think that change would also apply to other places where you're referencing `ApiTutorial` in this PR.

File: src/hooks/bookmarks/use-is-bookmarked.ts
Comment (kendallstrautman) [reply src/hooks/bookmarks/use-is-bookmarked.ts:20]: Right, for that use case I figured that the bookmark component itself wouldn't call the `isBookmarked` hook, and that data would be passed down from higher. 

But just based on the naming of this hook, I would expect it to query just a single bookmark.

File: src/hooks/bookmarks/use-bookmarks-by-tutorial-ids.ts
Comment (ashleemboyer) [reply src/hooks/bookmarks/use-bookmarks-by-tutorial-ids.ts:18]: The reason I didn't go with that idea right away is because it requires invoking this hook on every view we have TutorialCards. So far, that means product landing pages, tutorial landing pages, collection views, and install landing pages. The cache is being leveraged correctly here too! Even across several pages, all bookmarks are only being fetched one time. I think your idea is a good optimization to try out in a separate PR since it'll require refactors in a quite a few places and this PR is already doing a lot of refactoring.

File: src/hooks/bookmarks/use-is-bookmarked.ts
Comment (ashleemboyer) [reply src/hooks/bookmarks/use-is-bookmarked.ts:20]: While we could pass data down, one of the primary purposes of using the hooks is so we can avoid props drilling. The hook does return the value one single bookmark, it just makes sure the data it needs to do so is already present. And reduces the number of fetches made as a side effect.

Comment (kendallstrautman) [reply src/hooks/bookmarks/use-is-bookmarked.ts:20]: Well the tutorial meta only cares about one tutorial, not multiple. So the cards are where the multiple fetches come into play. Is there a way to prime a global cache in a way, such that if we make a call for a particular tutorial on say a product landing page, that data would be cached for the tutorial and collection page views? 

Or is this approach what that is intended to do in a way?

File: src/components/bookmark-button/index.tsx
Comment (ashleemboyer) [reply src/components/bookmark-button/index.tsx:13]: I already had to deal with managing conflicts on this branch after spending many hours on it, so I'd appreciate it ❤️

File: src/components/tutorial-card/types.ts
Comment (ashleemboyer) [reply src/components/tutorial-card/types.ts]: Which places? Many of them are valid uses of that interface

Comment (ashleemboyer) [reply src/components/tutorial-card/types.ts]: Updated in https://github.com/hashicorp/dev-portal/pull/838/commits/54013f9f393855d971bede11304c57b6b6da75e2!

File: src/hooks/bookmarks/use-is-bookmarked.ts
Comment (ashleemboyer) [reply src/hooks/bookmarks/use-is-bookmarked.ts:20]: The query priming approach wasn't working right away and needs some more time to explore, so that's another reason why optimizing this hook has been deferred for now. Once the individual query `bookmark` priming is worked out, then we can step `useIsBookmarked` toward no longer fetching all bookmarks to "prime" itself.

Comment (kendallstrautman) [reply src/hooks/bookmarks/use-is-bookmarked.ts:20]: In dev mode, I'm seeing that every time a bookmark is created or deleted the 'bookmarks' endpoint is being called with this set-up. Having all bookmarks called when just one needs to be updated doesn't seem like the behavior we want.

What was happening with the individual priming? Should we wait to figure that out before finishing up these hooks?

Comment (ashleemboyer) [reply src/hooks/bookmarks/use-is-bookmarked.ts:20]: Sorry if I wasn't clear! I want to overall optimize and also resolve the query priming in a separate PR since everything is working as needed here.

File: src/views/profile-view/index.tsx
Comment (kendallstrautman) [inline src/views/profile-view/index.tsx:24]: We will still need the user data for this view. Is there a reason you removed it?

Comment (kendallstrautman) [reply src/views/profile-view/index.tsx:24]: Actually let me check the designs again, maybe im getting wires crossed with Learn

**Review by kendallstrautman — APPROVED** [review]
Approving with the knowledge that optimization with the way were fetching data based on the threads will occur on follow-up PRs 👌

**Review by BRKalow — COMMENTED** [review]
Exciting stuff! There are a few things I think we want to adjust here in how we're typing the query hooks, namely relying on the underlying `useQuery` types for inferring the type of `data` instead of overwriting that declaration.

We can also simplify the `useIsBookmarked` hook to select data directly from the result of `useAllBookmarks`, if that is the preferred implementation.

As we start to introduce data fetching, let's also consider how wiring up these components will impact our ability to test and document them. If we add queries to components we render in swingset, we'll likely need to setup the query plumbing there as well. Sometimes having a purely presentational component can be a valuable separation!

File: src/components/tutorial-card/types.ts
Comment (ashleemboyer) [reply src/components/tutorial-card/types.ts]: I undid these changes, so marking resolved.

File: src/components/bookmark-button/index.tsx
Comment (ashleemboyer) [reply src/components/bookmark-button/index.tsx]: Undid these changes, marking resolved.

File: src/hooks/bookmarks/use-all-bookmarks.ts
Comment (ashleemboyer) [reply src/hooks/bookmarks/use-all-bookmarks.ts:6]: This is mainly just renaming `data` to `bookmarks` so the property has a clearer name that can make the hook consumption a little clearer. I can update it to reference `data` though, rather than remove + set our own.

File: src/hooks/bookmarks/use-all-bookmarks.ts
Comment (ashleemboyer) [reply src/hooks/bookmarks/use-all-bookmarks.ts:19]: It's still being inferred, this just renamed the `data` property to `bookmarks`.

File: src/hooks/bookmarks/use-bookmark-mutations.ts
Comment (ashleemboyer) [reply src/hooks/bookmarks/use-bookmark-mutations.ts:40]: I can make this update.

File: src/hooks/bookmarks/use-bookmarks-by-tutorial-ids.ts
Comment (ashleemboyer) [reply src/hooks/bookmarks/use-bookmarks-by-tutorial-ids.ts:28]: Ah good call, I forgot to come back to this. I paused on thinking about this query key because it's not extremely straightforward. We want to make sure this query gets invalidated on add/remove and I wasn't 100% sure (but will test) if the query you're proposing would get invalidated just based on the `"bookmarks"` portion.

File: src/hooks/bookmarks/use-is-bookmarked.ts
Comment (ashleemboyer) [reply src/hooks/bookmarks/use-is-bookmarked.ts:20]: I don't think the `select` option is what we want, because it'd cause the `find` to be invoked every time the hook renders. I wrapped the `find` in a `useQuery` so that wouldn't happen, and the value would be cached and accessible anywhere.

But I'll go back to the query priming and see if I can get that working as we want it to.

Comment (BRKalow) [reply src/hooks/bookmarks/use-is-bookmarked.ts:20]: I don't expect that the usage of `find` here will have any meaningful performance impact in practice. If we were noticing performance issues, the selector can also be optimized to ensure it only runs when the underlying data changes, per the linked post!

I can see some benefit to the unique queries per bookmark though, considering we'll need to invalidate them on mutations. 💭

File: src/hooks/bookmarks/use-all-bookmarks.ts
Comment (BRKalow) [inline src/hooks/bookmarks/use-all-bookmarks.ts:36]: Nicee!

**Review by BRKalow — APPROVED** [review]
👏

**Pull request merged**
